hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/prehandle/CompoundSignatureVerificationFuture.java

```diff
@@ -0,0 +1,218 @@
+ *
+ * <p>For example, if an instance is created with a {@link KeyList} based {@link Key}, then when {@link #get()} or
+ * {@link #get(long, TimeUnit)} are called, it will return a {@link SignatureVerification} where the
+ * {@link SignatureVerification#key()} is that key, and {@link SignatureVerification#passed()} is true if,
+ * and only if, every {@link SignatureVerification} in the List has also passed.
+ *
+ * <p>If an instance is created with a {@link ThresholdKey} based {@link Key}, then when {@link #get()} or
+ * {@link #get(long, TimeUnit)} are called, it will return a {@link SignatureVerification} where the
+ * {@link SignatureVerification#key()} is that key, and {@link SignatureVerification#passed()} is true if,
+ * and only if, it did not encounter {@link #numCanFail} failing verifications in the list.
+ */
+final class CompoundSignatureVerificationFuture implements SignatureVerificationFuture {
+    /**
+     * The Key we verified. This will *never* be null, because we would not have attempted signature verification
+     * without having a key. If an EVM address was used, we would have already extracted the key, so it can be
+     * provided here.
+     */
+    private final Key key;
+    /**
+     * Optional: used only with hollow accounts, the EVM address associated with the hollow account. The corresponding
+     * {@link #key} was extracted from the transaction (either because it was a "full" prefix on the signature map,
+     * or because we used an "ecrecover" like process to extract the key from the signature and signed bytes).
+     */
+    private final Account hollowAccount;
+    /**
+     * The list of {@link Future<TransactionSignature>}s. This list cannot be empty or contain null.
+     */
+    private final List<Future<SignatureVerification>> futures;
+    /**
+     * Whether *this* future has been canceled. Used for properly implementing {@link Future} semantics.
+     */
+    private boolean canceled = false;
+    /**
+     * The number of futures with verifications that can fail (passed == false) before the {@link SignatureVerification}
+     * created by this instance will have failed.
+     */
+    private final int numCanFail;
+
+    /**
+     * Create a new instance.
+     *
+     * @param key The key associated with this sig check. Cannot be null.
+     * @param hollowAccount The hollow account, if any
+     * @param futures The {@link TransactionSignature}s, from which the pass/fail status of the
+     * {@link SignatureVerification} is derived. This list must contain at least one element.
+     */
+    CompoundSignatureVerificationFuture(
+            @NonNull final Key key,
+            @Nullable final Account hollowAccount,
+            @NonNull final List<Future<SignatureVerification>> futures,
+            final int numCanFail) {
+        this.key = requireNonNull(key);
+        this.hollowAccount = hollowAccount;
+        this.futures = requireNonNull(futures);
```

Comment:
Maybe we should copy the list to an immutable list. Otherwise the `if (futures.isEmpty()) {` check at the end of the constructor is worthless since the list could be mutated from the outside.

Reply:
We could, and I definitely appreciate the safety argument. My only concern is garbage generation. The fact that `CompoundSignatureVerificationFuture` is package private and only used by `PreHandleResult` in one place where `PreHandleResult` creates the array list as a local variable in the function makes me think this level of (very very good) paranoia is maybe not needed in this case. I myself am very paranoid about code, and I like the idea of being very defensive. But I worry about all the object allocation too. I wish there was a better way to prove immutability of the list and throw an exception if it was ever a mutable list.